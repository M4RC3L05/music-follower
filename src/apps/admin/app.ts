import config from "config";
import Koa from "koa";
import bodyParser from "koa-bodyparser";
import render from "koa-ejs";
import qs from "koa-qs";
import session from "koa-session";
import koaStatic from "koa-static";
import path from "node:path";
import url from "node:url";

import router from "#src/apps/admin/router.js";

export const app = () => {
  const koa = new Koa();
  koa.keys = config.get("apps.admin.keys");

  qs(koa);
  render(koa, {
    root: path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "views"),
    layout: "main",
    viewExt: "ejs",
    cache: process.env.NODE_ENV === "production",
    debug: false,
  });

  koa.use(session({ ...config.get("apps.admin.session") }, koa));
  koa.use(bodyParser());
  koa.use(koaStatic(path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "../../../public")));
  koa.use(router.middleware());
  koa.use(router.allowedMethods());

  return koa;
};
