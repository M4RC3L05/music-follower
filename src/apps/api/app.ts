import { type Database } from "@leafac/sqlite";
import config from "config";
import { Hono } from "hono";
import { basicAuth } from "hono/basic-auth";
import { cors } from "hono/cors";
import { errorMappers } from "#src/errors/mod.js";
import { errorMapper, requestLifeCycle } from "#src/middlewares/mod.js";
import { router } from "./router.js";

export type Api = ReturnType<typeof makeApp>;

export const makeApp = ({ database }: { database: Database }) => {
  const app = new Hono();

  app.use("*", (c, next) => {
    c.set("database", database);

    return next();
  });
  app.use("*", requestLifeCycle);
  app.use("*", cors());
  app.use(
    "*",
    basicAuth({
      username: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .name,
      password: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .pass,
    }),
  );

  app.onError(
    errorMapper({
      mappers: [errorMappers.validationErrorMapper],
      defaultMapper: errorMappers.defaultErrorMapper,
    }),
  );

  return app.route("/api", router());
};
