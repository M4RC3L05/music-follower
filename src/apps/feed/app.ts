import { App } from "@m4rc3l05/sss";

import { feedMiddleware } from "./middlewares/feed-middleware.js";
import requestLifeCycle from "#src/middlewares/request-lifecycle.js";

export const makeApp = () => {
  const app = new App();

  app.use(requestLifeCycle);
  app.use(feedMiddleware);

  return app;
};
