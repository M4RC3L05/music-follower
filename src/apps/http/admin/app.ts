import Koa from "koa";
import basicAuth from "koa-basic-auth";
import config from "config";
import koaStatic from "koa-static";

export const app = () => {
  const koa = new Koa();

  koa.use(basicAuth({ ...config.get("apps.admin.basicAuth") }));
  koa.use(koaStatic("./src/apps/http/admin/public"));

  return koa;
};
