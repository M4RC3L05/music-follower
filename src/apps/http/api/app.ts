import Koa from "koa";
import bodyParser from "koa-bodyparser";
import cors from "@koa/cors";

import * as errorMappers from "#src/common/errors/mappers/mod.js";
import { errorMapper, requestLifeCycle } from "#src/common/middlewares/mod.js";
import logger from "#src/common/clients/logger.js";
import router from "./router.js";

export const app = () => {
  const app = new Koa();

  app.use(requestLifeCycle({ loggerFactory: logger }));
  app.use(
    errorMapper({
      loggerFactory: logger,
      defaultMapper: errorMappers.defaultErrorMapper,
      mappers: [errorMappers.validationErrorMapper],
    }),
  );
  app.use(cors());
  // App.use(basicAuth({ ...config.get("apps.api.basicAuth") }));
  app.use(bodyParser());
  app.use(router.routes());
  app.use(router.allowedMethods());

  return app;
};
