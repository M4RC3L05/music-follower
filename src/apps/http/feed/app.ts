import Koa from "koa";
import favicon from "koa-favicon";

import { feedMiddleware } from "#src/apps/http/feed/middlewares/feed-middleware.js";

export const app = () => {
  const koa = new Koa();

  koa.use(favicon("./src/apps/http/feed/favicon.ico"));
  koa.use(feedMiddleware);

  return koa;
};
