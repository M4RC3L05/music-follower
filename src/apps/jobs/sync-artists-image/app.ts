import { makeLogger } from "#src/common/logger/mod.ts";
import type { Artist } from "#src/database/types/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { delay } from "@std/async";
import { getArtistById } from "#src/remote/itunes/requests/mod.ts";
import { getArtistImage } from "#src/remote/apple-music/requests/mod.ts";

const log = makeLogger("sync-artists-image-job");

const runner = async ({
  db,
  abort,
}: { db: CustomDatabase; abort: AbortSignal }) => {
  if (abort.aborted) {
    return;
  }

  log.info("Begin artist image sync");

  const artists = db.prepare(`
    select *, row_number() over (order by id) as "index", ti."totalItems" as "totalItems"
    from
      artists,
      (select count(id) as "totalItems" from artists) as ti
    order by id;
  `).iter() as IterableIterator<Artist & { index: number; totalItems: number }>;

  for (const artist of artists) {
    const isLastArtist = artist.index >= artist.totalItems;

    if (abort.aborted) {
      break;
    }

    log.info(
      `Sync image from "${artist.name}" at ${artist.index} of ${artist.totalItems}`,
    );

    let imageUrl: string;

    try {
      const artistRemote = await getArtistById(artist.id, abort);

      if (!artistRemote) {
        throw new Error(`No remote artist found with id ${artist.id}`);
      }

      imageUrl = await getArtistImage(
        artistRemote.artistLinkUrl,
        abort,
      );
    } catch (error: unknown) {
      log.error("Something went wrong fetching artist image", { error });

      if (!isLastArtist) {
        log.info("Waiting 5 seconds before processing next artist");

        await delay(5000, { signal: abort }).catch(() => {});
      }

      continue;
    }

    if (abort.aborted) {
      break;
    }

    try {
      log.info("Updating artist image", { id: artist.id });

      db.sql`update artists set image = ${imageUrl} where id = ${artist.id}`;

      log.info("Artist image updated", { id: artist.id });
    } catch (error: unknown) {
      log.error("Something wrong ocurred while updating artist image", {
        error,
      });
    }

    if (!isLastArtist) {
      log.info("Waiting 5 seconds before processing next artist");

      await delay(5000, { signal: abort }).catch(() => {});
    }
  }

  log.info("Artist image sync ended");
};

export default runner;
