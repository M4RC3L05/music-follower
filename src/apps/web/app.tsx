import { serveStatic } from "@hono/hono/deno";
import config from "config";
import { type ContextVariableMap, Hono } from "@hono/hono";
import {
  MemoryStore,
  type Session,
  sessionMiddleware,
} from "@jcs224/hono-sessions";
import { csrf } from "@hono/hono/csrf";
import { jsxRenderer } from "@hono/hono/jsx-renderer";
import { HTTPException } from "@hono/hono/http-exception";
import { secureHeaders } from "@hono/hono/secure-headers";
import { makeLogger } from "#src/common/logger/mod.ts";
import { router } from "#src/apps/web/routes/mod.ts";
import { MainLayout } from "#src/apps/web/views/common/layouts/main.tsx";
import { serviceRegister } from "#src/middlewares/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import { PagesErrorPage } from "./views/pages/pages/error.tsx";
import type { MiddlewareHandler } from "@hono/hono";
import { STATUS_TEXT, type StatusCode } from "@std/http";
import type {
  SessionFlashFormErrors,
  SessionFlashMessages,
} from "#src/apps/web/types.ts";
import { MultipartParseError } from "@mjackson/multipart-parser";
import { errors } from "@vinejs/vine";
import { feedRoutes } from "#src/apps/web/routes/feed/mod.ts";

const log = makeLogger("admin");

declare module "@hono/hono" {
  interface ContextVariableMap {
    database: CustomDatabase;
    shutdown: AbortSignal;
    session: Session<
      {
        flashMessages?: SessionFlashMessages;
        flashFormErrors?: SessionFlashFormErrors;
      }
    >;
  }

  interface ContextRenderer {
    (
      content: string | Promise<string>,
      props?: { error?: boolean },
    ): Response;
  }
}

export const makeApp = (deps: Partial<ContextVariableMap>) => {
  const app = new Hono();

  app.onError((error, c) => {
    log.error("Something went wrong!", { error });

    if (error instanceof HTTPException) {
      if (error.res) {
        error.res.headers.forEach((value, key) => {
          c.header(key, value);
        });
      }
    }

    if ((error instanceof HTTPException && error.status === 422)) {
      c.get("session").flash("flashMessages", { error: [error.message] });

      if (error.cause) {
        c.get("session").flash(
          "flashFormErrors",
          error.cause as SessionFlashFormErrors,
        );
      }

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof errors.E_VALIDATION_ERROR) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      c.get("session").flash(
        "flashFormErrors",
        (error.messages as { message: string; field: string }[]).reduce(
          (acc, curr) => {
            const errMessages = acc[curr.field] ?? [];
            errMessages.push(curr.message);
            acc[curr.field] = errMessages;
            return acc;
          },
          {} as Record<string, string[]>,
        ),
      );

      return c.redirect(c.req.header("Referer") ?? "/");
    }

    if (error instanceof MultipartParseError) {
      c.get("session").flash("flashMessages", { error: [error.message] });
      return c.redirect(c.req.header("Referer") ?? "/");
    }

    const errorMessage = error instanceof HTTPException
      ? (error.message.trim().length > 0 ? error.message : undefined) ??
        STATUS_TEXT[error.status as StatusCode] ??
        "Something broke..."
      : "Something broke...";
    c.status(error instanceof HTTPException ? error.status : 500);

    return c.render(<PagesErrorPage errorMessage={errorMessage} />, {
      error: true,
    });
  });

  app.notFound((c) => {
    c.status(404);
    return c.render(<PagesErrorPage errorMessage="Not found" />, {
      error: true,
    });
  });

  app.use("*", async (c, next) => {
    const nocache = () => {
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      c.header("cache-control", "no-cache, no-store, must-revalidate");
      c.header("pragma", "no-cache");
      c.header("expires", "0");
    };

    try {
      await next();

      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }
    } catch (error) {
      if (
        !(c.req.path.includes("/favicon") || c.req.path.includes("/public") ||
          c.req.path.includes("/deps"))
      ) {
        nocache();
      } else {
        c.header("Cache-Control", `max-age=${60 * 60 * 24}`);
      }

      throw error;
    }
  });
  app.use("*", serviceRegister(deps));
  app.use(
    "*",
    secureHeaders({ referrerPolicy: "strict-origin-when-cross-origin" }),
  );

  app.route("/feeds", feedRoutes());
  app.get("/favicon.ico", serveStatic({ root: "./src/apps/web/public" }));
  app.get(
    "/public/*",
    serveStatic({
      root: "./src/apps/web/public",
      rewriteRequestPath: (path) => path.replace("/public", ""),
    }),
  );
  app.get(
    "/deps/*",
    serveStatic({
      root: "./node_modules",
      rewriteRequestPath: (path) => path.replace("/deps", ""),
    }),
  );

  app.use(
    "*",
    sessionMiddleware({
      store: new MemoryStore(),
      encryptionKey: "foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar",
      expireAfterSeconds: 60 * 15,
      sessionCookieName: "sid",
      cookieOptions: {
        sameSite: "strict",
        path: "/",
        httpOnly: true,
        secure: true,
      },
    }) as unknown as MiddlewareHandler,
  );

  app.use(
    "*",
    jsxRenderer(({ children, error = false }, c) => {
      const flashMessages = c.get("session").get("flashMessages");

      return (
        <MainLayout
          path={c.req.path}
          flashMessages={flashMessages ?? undefined}
          enableNavbar={!error}
        >
          {children}
        </MainLayout>
      );
    }, {
      docType: true,
      stream: true,
    }),
  );

  app.use(
    "*",
    csrf({
      origin: (origin) => new RegExp(config.get("csrf.origin")).test(origin),
    }),
  );

  if (Deno.env.get("ENV") === "test") {
    app.get("/test/flash-messages", (c) => {
      c.get("session").flash("flashMessages", {
        error: ["error", "error2"],
        info: ["info"],
        success: ["success"],
        warning: ["warning"],
      });

      return c.render("");
    });

    app.get("/test/error", (c) => {
      const { type } = c.req.query();

      if (type === "http") {
        throw new HTTPException(400, {
          res: new Response(null, { headers: { "x-foo": "bar" } }),
        });
      }

      if (type === "http422") {
        throw new HTTPException(422, {
          message: "foo",
          cause: { foo: ["yes"] },
        });
      }

      if (type === "valierr") {
        const e = new errors.E_VALIDATION_ERROR([{
          message: "lorem ipsum",
          field: "foo",
        }, {
          message: "lorem lorem",
          field: "bar",
        }]);

        e.message = "foo";

        throw e;
      }

      if (type === "multparterr") {
        throw new MultipartParseError("foo");
      }

      throw new Error("foo");
    });
  }

  return app.route("/", router());
};
