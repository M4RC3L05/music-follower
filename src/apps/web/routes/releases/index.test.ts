import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "@std/testing/bdd";
import { type CustomDatabase, makeDatabase } from "#src/database/mod.ts";
import { testDbUtils, testUtils } from "#src/common/utils/mod.ts";
import { makeApp } from "#src/apps/web/app.tsx";
import type { Hono } from "@hono/hono";
import * as testFixtures from "#src/common/test-fixtures/mod.ts";
import { assertEquals } from "@std/assert";
import type { Release } from "#src/database/types/mod.ts";
import type { JSDOM } from "jsdom";
import { MemoryStore } from "@jcs224/hono-sessions";

let app: Hono;
let db: CustomDatabase;
let memStore: MemoryStore;

beforeAll(async () => {
  db = makeDatabase();
  await testDbUtils.runMigrations(db);

  memStore = new MemoryStore();
  app = makeApp({ database: db, sessioStore: memStore });
});

beforeEach(() => {
  db.exec("delete from releases");
  db.exec("delete from accounts");
  testFixtures.loadAccount(db);
  // deno-lint-ignore ban-ts-comment
  // @ts-ignore
  memStore.data.clear();
});

afterAll(() => {
  db.close();
});

describe("GET /releases", () => {
  describe("snapshot", () => {
    it("should display without releases", async (t) => {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request(`/releases`, {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      await testUtils.expectHtmlSnapshot(t, dom, "body");
    });

    it("should display with releases", async (t) => {
      testFixtures.loadRelease(db, { id: 1 });
      testFixtures.loadRelease(db, { id: 2 });

      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request(`/releases`, {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      await testUtils.expectHtmlSnapshot(t, dom, "body");
    });
  });

  it("should redirect to login page if not authenticated", async () => {
    const res = await app.request("/releases");

    assertEquals(res.status, 302);
    assertEquals(res.headers.get("location"), "/auth/login");
  });

  it("should mark releases item in navbar as active", async () => {
    const auth = await testUtils.authenticateRequest(app);
    const res = await app.request("/releases", {
      headers: { cookie: auth.sid },
    });
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 200);
    testUtils.assertSee(
      dom,
      '<a class="nav-link active" href="/releases">Releases</a>',
      "nav.navbar",
    );
  });

  it("should render releases index page without releases", async () => {
    const auth = await testUtils.authenticateRequest(app);
    const res = await app.request("/releases", {
      headers: { cookie: auth.sid },
    });
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 200);
    testUtils.assertNodeExists(dom, "header", "body");
    testUtils.assertSeeInOrder(
      dom,
      [
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;page=0"><i class="bi bi-chevron-double-left"></i> Start</a>',
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;page=0"><i class="bi bi-chevron-left"></i> Previous</a>',
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;page=0">Next <i class="bi bi-chevron-right"></i></a>',
        '<a class="btn btn-sm btn-outline-primary" href="/releases?limit=12&amp;page=0">End <i class="bi bi-chevron-double-right"></i></a>',
      ],
      "body header",
    );
    testUtils.assertNodeExists(dom, "main", "body");
    testUtils.assertSee(
      dom,
      '<p class="text-center">No releases to show</p>',
      "body main",
    );
  });

  it("should render orm errors if invalid search params provided", async () => {
    const auth = await testUtils.authenticateRequest(app);
    const res = await app.request("/releases?hidden=b", {
      headers: { referer: "http://localhost/releases", cookie: auth.sid },
    });

    assertEquals(res.status, 302);
    assertEquals(res.headers.get("location"), "http://localhost/releases");

    const res2 = await app.request("/releases", {
      headers: { cookie: auth.sid },
    });
    const dom = await testUtils.getDom(res2);

    testUtils.assertSee(
      dom,
      '<div class="invalid-feedback"><p key="0">The selected hidden is invalid',
      "body header form",
    );
  });

  it("should render releases index with links with query params", async () => {
    const auth = await testUtils.authenticateRequest(app);
    const res = await app.request("/releases?q=b", {
      headers: { cookie: auth.sid },
    });
    const dom = await testUtils.getDom(res);

    assertEquals(res.status, 200);
    testUtils.assertNodeExists(dom, "header", "body");
    testUtils.assertSeeInOrder(
      dom,
      [
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;q=b&amp;page=0"><i class="bi bi-chevron-double-left"></i> Start</a>',
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;q=b&amp;page=0"><i class="bi bi-chevron-left"></i> Previous</a>',
        '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=12&amp;q=b&amp;page=0">Next <i class="bi bi-chevron-right"></i></a>',
        '<a class="btn btn-sm btn-outline-primary" href="/releases?limit=12&amp;q=b&amp;page=0">End <i class="bi bi-chevron-double-right"></i></a>',
      ],
      "body header",
    );
    testUtils.assertNodeExists(dom, "main", "body");
    testUtils.assertSee(
      dom,
      '<p class="text-center">No releases to show</p>',
      "body main",
    );
  });

  it("should render releases index page with correct pagination with releases", async () => {
    const releaseOne = testFixtures.loadRelease(db, {
      id: 1,
      name: "1foo",
      releasedAt: new Date(Date.now() + 60 * 1000).toISOString(),
    });
    const releaseTwo = testFixtures.loadRelease(db, {
      id: 2,
      name: "2bar",
      releasedAt: new Date(100).toISOString(),
    });
    const releaseThree = testFixtures.loadRelease(db, {
      id: 3,
      name: "3biz",
      releasedAt: new Date(10).toISOString(),
    });
    const releaseFour = testFixtures.loadRelease(db, {
      id: 4,
      name: "4buz",
      releasedAt: new Date(1).toISOString(),
    });
    const releaseFive = testFixtures.loadRelease(db, {
      id: 5,
      name: "5bax",
      releasedAt: new Date(0).toISOString(),
    });

    const checkReleaseInPage = (dom: JSDOM, release: Release) => {
      testUtils.assertNodeExists(
        dom,
        `main img[src="${release.coverUrl}"]`,
        "body",
      );

      testUtils.assertSee(
        dom,
        `<a class="btn btn-outline-primary" href="/releases/${release.id}/${release.type}">`,
        "body main",
      );

      let txt = [];

      if (release.id === 1) txt.push("To be released");

      txt = [
        release.type,
        release.name,
        release.artistName,
        new Date(release.releasedAt).toISOString(),
      ];

      testUtils.assertSeeTextInOrder(dom, txt, "body main");
    };

    {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request("/releases?page=0&limit=2", {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      testUtils.assertNodeExists(dom, "header", "body");
      testUtils.assertSeeInOrder(
        dom,
        [
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=0"><i class="bi bi-chevron-double-left"></i> Start</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=0"><i class="bi bi-chevron-left"></i> Previous</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=1">Next <i class="bi bi-chevron-right"></i></a>',
          '<a class="btn btn-sm btn-outline-primary" href="/releases?limit=2&amp;page=2">End <i class="bi bi-chevron-double-right"></i></a>',
        ],
        "body header",
      );
      testUtils.assertNodeExists(dom, "main", "body");
      testUtils.assertSeeTextInOrder(
        dom,
        [releaseOne.name, releaseTwo.name],
        "body main",
      );
      checkReleaseInPage(dom, releaseOne);
      checkReleaseInPage(dom, releaseTwo);
    }

    {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request("/releases?page=1&limit=2", {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      testUtils.assertNodeExists(dom, "header", "body");
      testUtils.assertSeeInOrder(
        dom,
        [
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=0"><i class="bi bi-chevron-double-left"></i> Start</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=0"><i class="bi bi-chevron-left"></i> Previous</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=2">Next <i class="bi bi-chevron-right"></i></a>',
          '<a class="btn btn-sm btn-outline-primary" href="/releases?limit=2&amp;page=2">End <i class="bi bi-chevron-double-right"></i></a>',
        ],
        "body header",
      );
      testUtils.assertNodeExists(dom, "main", "body");
      testUtils.assertSeeTextInOrder(
        dom,
        [releaseThree.name, releaseFour.name],
        "body main",
      );
      checkReleaseInPage(dom, releaseThree);
      checkReleaseInPage(dom, releaseFour);
    }

    {
      const auth = await testUtils.authenticateRequest(app);
      const res = await app.request("/releases?page=2&limit=2", {
        headers: { cookie: auth.sid },
      });
      const dom = await testUtils.getDom(res);

      assertEquals(res.status, 200);
      testUtils.assertNodeExists(dom, "header", "body");
      testUtils.assertSeeInOrder(
        dom,
        [
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=0"><i class="bi bi-chevron-double-left"></i> Start</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=1"><i class="bi bi-chevron-left"></i> Previous</a>',
          '<a class="btn btn-sm btn-outline-primary me-2" href="/releases?limit=2&amp;page=2">Next <i class="bi bi-chevron-right"></i></a>',
          '<a class="btn btn-sm btn-outline-primary" href="/releases?limit=2&amp;page=2">End <i class="bi bi-chevron-double-right"></i></a>',
        ],
        "body header",
      );
      testUtils.assertNodeExists(dom, "main", "body");
      testUtils.assertSeeText(dom, releaseFive.name, "body main");
      checkReleaseInPage(dom, releaseFive);
    }
  });
});
