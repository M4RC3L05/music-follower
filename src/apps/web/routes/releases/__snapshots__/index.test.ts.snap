export const snapshot = {};

snapshot[`global > GET /releases > snapshot > should display without releases 1`] = `
'
<body>
  <nav class="navbar navbar-expand-lg bg-body-tertiary">
    <div class="container">
      <a
        class="navbar-brand"
        href="/"
      >
        Music Follower
      </a>
      <button
        class="navbar-toggler"
        type="button"
        data-bs-toggle="collapse"
        data-bs-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span class="navbar-toggler-icon">
        </span>
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul class="navbar-nav me-auto mb-2 mb-lg-0">
          <li class="nav-item">
            <a
              class="nav-link "
              aria-current="page"
              href="/"
            >
              Home
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link "
              href="/artists"
            >
              Artists
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link active"
              href="/releases"
            >
              Releases
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <header
    class="mx-auto my-3 position-sticky z-2 mb-5"
    style="top:1rem"
  >
    <div class="container">
      <div class="row">
        <div class="col">
          <div class="card px-3 py-4">
            <div calss="card-body">
              <form
                class="row g-2 justify-content-center mb-3"
                action="/releases"
                method="get"
              >
                <div class="col-auto">
                  <input
                    class="form-control "
                    type="text"
                    name="q"
                    placeholder="Search release"
                    value
                  >
                </div>
                <div class="col-auto">
                  <select
                    name="hidden"
                    class="form-control "
                  >
                    <option value="none">
                      Filter hidden mode
                    </option>
                    <option value="admin">
                      Admin
                    </option>
                    <option value="feed">
                      Feed
                    </option>
                  </select>
                </div>
                <div class="col-auto">
                  <button
                    type="submit"
                    class="btn btn-outline-primary"
                  >
                    Search
                    <i class="bi bi-search">
                    </i>
                  </button>
                </div>
              </form>
              <div class="d-flex justify-content-center mb-2">
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  <i class="bi bi-chevron-double-left">
                  </i>
                  Start
                </a>
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  <i class="bi bi-chevron-left">
                  </i>
                  Previous
                </a>
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  Next
                  <i class="bi bi-chevron-right">
                  </i>
                </a>
                <a
                  class="btn btn-sm btn-outline-primary"
                  href="/releases?limit=12&amp;page=0"
                >
                  End
                  <i class="bi bi-chevron-double-right">
                  </i>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <main>
    <div class="container">
      <div class="row">
        <p class="text-center">
          No releases to show
        </p>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;

snapshot[`global > GET /releases > snapshot > should display with releases 1`] = `
'
<body>
  <nav class="navbar navbar-expand-lg bg-body-tertiary">
    <div class="container">
      <a
        class="navbar-brand"
        href="/"
      >
        Music Follower
      </a>
      <button
        class="navbar-toggler"
        type="button"
        data-bs-toggle="collapse"
        data-bs-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span class="navbar-toggler-icon">
        </span>
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul class="navbar-nav me-auto mb-2 mb-lg-0">
          <li class="nav-item">
            <a
              class="nav-link "
              aria-current="page"
              href="/"
            >
              Home
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link "
              href="/artists"
            >
              Artists
            </a>
          </li>
          <li class="nav-item">
            <a
              class="nav-link active"
              href="/releases"
            >
              Releases
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <header
    class="mx-auto my-3 position-sticky z-2 mb-5"
    style="top:1rem"
  >
    <div class="container">
      <div class="row">
        <div class="col">
          <div class="card px-3 py-4">
            <div calss="card-body">
              <form
                class="row g-2 justify-content-center mb-3"
                action="/releases"
                method="get"
              >
                <div class="col-auto">
                  <input
                    class="form-control "
                    type="text"
                    name="q"
                    placeholder="Search release"
                    value
                  >
                </div>
                <div class="col-auto">
                  <select
                    name="hidden"
                    class="form-control "
                  >
                    <option value="none">
                      Filter hidden mode
                    </option>
                    <option value="admin">
                      Admin
                    </option>
                    <option value="feed">
                      Feed
                    </option>
                  </select>
                </div>
                <div class="col-auto">
                  <button
                    type="submit"
                    class="btn btn-outline-primary"
                  >
                    Search
                    <i class="bi bi-search">
                    </i>
                  </button>
                </div>
              </form>
              <div class="d-flex justify-content-center mb-2">
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  <i class="bi bi-chevron-double-left">
                  </i>
                  Start
                </a>
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  <i class="bi bi-chevron-left">
                  </i>
                  Previous
                </a>
                <a
                  class="btn btn-sm btn-outline-primary me-2"
                  href="/releases?limit=12&amp;page=0"
                >
                  Next
                  <i class="bi bi-chevron-right">
                  </i>
                </a>
                <a
                  class="btn btn-sm btn-outline-primary"
                  href="/releases?limit=12&amp;page=0"
                >
                  End
                  <i class="bi bi-chevron-double-right">
                  </i>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <main>
    <div class="container">
      <div class="row">
        <div class="col-12">
          <section class="mb-4 text-center text-md-start d-block d-md-flex">
            <img
              class="img-fluid rounded me-md-5 mb-2 mb-md-0"
              src="http://example.com"
              style="aspect-ratio:1 / 1;max-width:250px"
            >
            <div>
              <span class="badge rounded-pill text-bg-primary mb-2">
                track
              </span>
              <h3>
                bar by foo
              </h3>
              <p>
                Released at 1970-01-01T00:00:00.000Z
              </p>
              <a
                class="btn btn-outline-primary"
                href="/releases/1/track"
              >
                More
              </a>
            </div>
          </section>
          <hr>
        </div>
        <div class="col-12">
          <section class="mb-4 text-center text-md-start d-block d-md-flex">
            <img
              class="img-fluid rounded me-md-5 mb-2 mb-md-0"
              src="http://example.com"
              style="aspect-ratio:1 / 1;max-width:250px"
            >
            <div>
              <span class="badge rounded-pill text-bg-primary mb-2">
                track
              </span>
              <h3>
                bar by foo
              </h3>
              <p>
                Released at 1970-01-01T00:00:00.000Z
              </p>
              <a
                class="btn btn-outline-primary"
                href="/releases/2/track"
              >
                More
              </a>
            </div>
          </section>
          <hr>
        </div>
      </div>
    </div>
  </main>
  <script
    type="module"
    src="/deps/bootstrap/dist/js/bootstrap.bundle.js"
  >
  </script>
</body>
'
`;
